faunadb/streams/client.py

```diff
@@ -0,0 +1,123 @@
+import threading
+from time import time
+
+try:
+    #python2
+    from urllib import urlencode
+except ImportError:
+    #python3
+    from urllib.parse import urlencode
+
+from hyper import HTTP20Connection
+from faunadb._json import to_json, parse_json_or_none
+from faunadb.request_result import RequestResult
+from .events import parse_stream_request_result_or_none, Error
+from .errors import StreamError
+
+VALID_FIELDS = {"ref", "ts", "diff", "old", "new", "action"}
+
+class Client(object):
+    pass
+
+class Connection(object):
+    """
+    The internal stream client connection interface.
+    This class handles the network side of a stream
+    subscription.
+
+    Current limitations:
+    Python requests module uses HTTP1; hyper is used for HTTP/2
+    """
+    def __init__(self, client, expression, options):
+        self._client = client
+        self.options = options
+        self.conn = None
+        self._fields = None
+        if isinstance(self.options, dict):
+            self._fields = self.options.get("fields", None)
+        elif hasattr(self.options, "fields"):
+            self._fields = self.options.field
+        if isinstance(self._fields, list):
+            union = set(self._fields).union(VALID_FIELDS)
+            if union != VALID_FIELDS:
+                raise Exception("Valid fields options are %s, provided %s."%(VALID_FIELDS, self._fields))
+        self._state = "idle"
+        self._query = expression
+        self._data = to_json(expression).encode()
+        try:
+            self.conn = HTTP20Connection(
+                    self._client.domain, port=self._client.port, enable_push=True)
+        except Exception as e:
+            raise StreamError(e)
+
+    def close(self):
+        """
+        Closes the stream subscription by aborting its underlying http request.
+        """
+        if self.conn is None:
+            raise StreamError('Cannot close inactive stream subscription.')
+        self.conn.close()
+        self._state = 'closed'
+
+    def error(self):
+        """
+        Retrieves the error from the event loop in an async implementation.
+        """
+        return self.error
+
+    def subscribe(self, on_event, blocking=False):
+        """Initiates the stream subscription."""
+        if self._state != "idle":
+            raise StreamError('Stream subscription already started.')
```

Comment:
Maybe we could have a unit test for this error.
WDYT?

Reply:
There is a test for this. [pull/165/files#diff-c7d25fbdfe94b27dca4d863a04f31fd9222cca8b19503976032b95f72f0aa923R131](https://github.com/fauna/faunadb-python/pull/165/files#diff-c7d25fbdfe94b27dca4d863a04f31fd9222cca8b19503976032b95f72f0aa923R131)